src/systems/wheel_slip/WheelSlip.cc

```diff
@@ -0,0 +1,346 @@
+    }
+
+    if (wheelElem->HasElement("wheel_radius"))
+    {
+      params.wheelRadius = wheelElem->Get<double>("wheel_radius");
+    }
+
+    auto link = Link(this->model.LinkByName(_ecm, linkName));
+    if (!link.Valid(_ecm))
+    {
+      ignerr << "Could not find link named [" << linkName
+            << "] in model [" << modelName << "]"
+            << std::endl;
+      continue;
+    }
+
+    auto collisions =
+        _ecm.ChildrenByComponents(link.Entity(), components::Collision());
+    if (collisions.empty() || collisions.size() != 1)
+    {
+      ignerr << "There should be 1 collision in link named [" << linkName
+            << "] in model [" << modelName << "]"
+            << ", but " << collisions.size() << " were found"
+            << std::endl;
+      continue;
+    }
+
+
+    auto collision = collisions.front();
+
+    params.collision = collision;
+
+    auto joints =
+        _ecm.ChildrenByComponents(model.Entity(), components::Joint(),
+                                  components::ChildLinkName(linkName));
+    if (joints.empty() || joints.size() != 1)
+    {
+      ignerr << "There should be 1 parent joint for link named [" << linkName
+            << "] in model [" << modelName << "]"
+            << ", but " << joints.size() << " were found"
+            << std::endl;
+      continue;
+    }
+
+    params.joint = joints.front();
+
+    if (params.wheelRadius <= 0)
+    {
+      ignerr << "Found wheel radius [" << params.wheelRadius
+            << "], which is not positive"
+            << " in link named [" << linkName
+            << "] in model [" << modelName << "]"
+            << std::endl;
+      continue;
+    }
+
+    if (params.wheelNormalForce <= 0)
+    {
+      ignerr << "Found wheel normal force [" << params.wheelNormalForce
+            << "], which is not positive"
+            << " in link named [" << linkName
+            << "] in model [" << modelName << "]"
+            << std::endl;
+      continue;
+    }
+
+    this->mapLinkSurfaceParams[link.Entity()] = params;
+  }
+
+  if (this->mapLinkSurfaceParams.empty())
+  {
+    ignerr << "No ODE links and surfaces found, plugin is disabled"
+           << std::endl;
+    return false;
+  }
+
+  return true;
+}
+
+/////////////////////////////////////////////////
+void WheelSlipPrivate::Update(EntityComponentManager &_ecm)
+{
+  for (const auto &linkSurface : this->mapLinkSurfaceParams)
+  {
+    const auto &params = linkSurface.second;
+
+    // get user-defined normal force constant
+    double force = params.wheelNormalForce;
+
+    auto joint = params.joint;
+
+    auto spinAngularVelocityComp =
+        _ecm.Component<components::JointVelocity>(joint);
+
+    if (!spinAngularVelocityComp || spinAngularVelocityComp->Data().empty())
+      continue;
+    double spinAngularVelocity = spinAngularVelocityComp->Data()[0];
+
+    // As discussed in WheelSlipPlugin.hh, the ODE slip1 and slip2
```

Comment:
I think we shouldn't speak in terms of ODE anymore :thinking:

Reply:
a1b6d997612965bd16cdd64dabb09585855e1bdf